librz/analysis/dwarf_process.c

```diff
@@ -3,237 +3,524 @@
+	if (btype) {
+		btype->type = NULL;
+	}
+	rz_type_base_type_free(btype);
+	return NULL;
+}
+
+/**
+ * \brief Parse and return the count of an array or 0 if not found/not defined
+ */
+static ut64 array_count_parse(Context *ctx, RzBinDwarfDie *die) {
+	if (!die->has_children) {
+		return 0;
+	}
+	RzPVector *children = die_children(die, ctx->dw);
+	if (!children) {
+		return 0;
+	}
+
+	void **it;
+	rz_pvector_foreach (children, it) {
+		RzBinDwarfDie *child_die = *it;
+		if (!(child_die->tag == DW_TAG_subrange_type)) {
+			continue;
+		}
+		RzBinDwarfAttr *value;
+		rz_vector_foreach(&child_die->attrs, value) {
+			switch (value->name) {
+			case DW_AT_upper_bound:
+			case DW_AT_count:
+				rz_pvector_free(children);
+				return value->uconstant + 1;
+			default:
+				break;
+			}
+		}
+	}
+	rz_pvector_free(children);
+	return 0;
+}
+
+/**
+ * Parse the die's DW_AT_type type or return a void type or NULL if \p type_idx == -1
+ *
+ * \param allow_void whether to return a void type instead of NULL if there is no type defined
+ */
+static RzType *type_parse_from_die_internal(Context *ctx, RzBinDwarfDie *die, bool allow_void, RZ_NULLABLE ut64 *size, RZ_NONNULL SetU *visited) {
+	RzBinDwarfAttr *attr = rz_bin_dwarf_die_get_attr(die, DW_AT_type);
+	if (!attr) {
+		if (!allow_void) {
+			return NULL;
+		}
+		return rz_type_identifier_of_base_type_str(ctx->analysis->typedb, "void");
+	}
+	return type_parse_from_offset_internal(ctx, attr->reference, size, visited);
+}
+
+/**
+ * \brief Recursively parses type entry of a certain offset and saves type size into *size
+ *
+ * \param ctx
+ * \param offset offset of the type entry
+ * \param size ptr to size of a type to fill up (can be NULL if unwanted)
+ * \return the parsed RzType or NULL on failure
+ */
+static RzType *type_parse_from_offset_internal(Context *ctx, ut64 offset,
+	RZ_NULLABLE ut64 *size, RZ_NONNULL SetU *visited) {
+	RzType *ret = ht_up_find(ctx->analysis->debug_info->type_by_offset, offset, NULL);
+	if (ret) {
+		return rz_type_clone(ret);
+	}
+
+	if (set_u_contains(visited, offset)) {
+		return NULL;
+	}
+	set_u_add(visited, offset);
+
+	RzBinDwarfDie *die = ht_up_find(ctx->dw->info->die_tbl, offset, NULL);
+	if (!die) {
+		return NULL;
+	}
+
+	// get size of first type DIE that has size
+	if (size && *size == 0) {
+		*size = die_bits_size(die);
+	}
+	switch (die->tag) {
+	// this should be recursive search for the type until you find base/user defined type
+	case DW_TAG_pointer_type:
+	case DW_TAG_reference_type: // C++ references are just pointers to us
+	case DW_TAG_rvalue_reference_type: {
+		RzType *pointee = type_parse_from_die_internal(ctx, die, true, size, visited);
+		if (!pointee) {
+			goto end;
+		}
+		ret = rz_type_pointer_of_type(ctx->analysis->typedb, pointee, false);
+		if (!ret) {
+			rz_type_free(pointee);
+			goto end;
+		}
+		break;
+	}
+	// We won't parse them as a complete type, because that will already be done
+	// so just a name now
+	case DW_TAG_typedef:
```

Comment:
I wonder if the approach similar to the C type parser could be used - see `state->forward` in `CParserState` in `librz/type/parser/` and `c_parser_forward_definition_store()`

Reply:
For example this function `RZ_OWN ParserTypePair *c_parser_new_typedef_naked_type(CParserState *state, RZ_NONNULL const char *name)`.

Reply:
And `RZ_OWN ParserTypePair *c_parser_new_typedef_forward_definition(CParserState *state, RZ_NONNULL const char *name)`